Frends.Community.RabbitMQ/RabbitMQTask.cs

```diff
@@ -12,88 +12,196 @@ namespace Frends.Community.RabbitMQ
+                }
+                
+                if (inputParams.WriteMessageCount <= 1)
+                {
+                    channel.BasicPublish(exchange:
+                        inputParams.ExchangeName,
+                        routingKey: inputParams.RoutingKey,
+                        basicProperties: basicProperties,
+                        body: inputParams.Data);
+                    return true;
+                }
 
+                return false;
+            }
+            finally
+            {
+               CloseConnection(channel, connection); 
             }
+        }
+        
+        /// <summary>
+        /// Writes messages into a queue with batch publish. All messages are under transaction. If one the message failes all messages will be rolled back.  
+        /// </summary>
+        /// <param name="inputParams"></param>
+        public static bool WriteBatchMessage([PropertyTab] WriteInputParams inputParams)
+        {
+            IConnection connection = CreateConnection(inputParams.HostName, inputParams.ConnectWithURI);
+            IModel channel = connection.CreateModel();
+            try
+            {
+                if (inputParams.Create)
+                {
+                    channel.QueueDeclare(queue: inputParams.QueueName,
+                        durable: inputParams.Durable,
+                        exclusive: false,
+                        autoDelete: false,
+                        arguments: null);
+                    channel.ConfirmSelect();
+                }
 
-            _channel.BasicPublish(exchange: inputParams.ExchangeName,
-                                    routingKey: inputParams.RoutingKey,
-                                    basicProperties: basicProperties,
-                                    body: inputParams.Data);
+                IBasicProperties basicProperties = null;
+
+                if (inputParams.Durable == true)
+                {
+                    basicProperties = channel.CreateBasicProperties();
+                    basicProperties.Persistent = true;
+                }
+            
+                // Add message into a memory based on producer write capability.
+                if (channel.MessageCount(inputParams.QueueName) <= inputParams.WriteMessageCount)
+                {
+                    CreateBasicPublishBatch(inputParams.ProcessExecutionId, channel).Add(
+                        exchange: inputParams.ExchangeName,
+                        routingKey: inputParams.RoutingKey,
+                        mandatory: true,
+                        properties: basicProperties,
+                        body: inputParams.Data);
+                    return false;
+                }
 
-            return true;
+                // Commit under transaction when all of the messages have been received for the producer.
+                if (channel.MessageCount(inputParams.QueueName) == inputParams.WriteMessageCount)
```

Comment:
Using message counts for specifying when to commit a message seems a bit error prone. It would be better to have a separate Task for committing the batch.